planning/behavior_path_planner/src/scene_module/utils/geometric_parallel_parking.cpp

```diff
@@ -68,83 +80,265 @@ PathWithLaneId GeometricParallelParking::getCurrentPath() const
 PathWithLaneId GeometricParallelParking::getFullPath() const
 {
   PathWithLaneId path{};
-  for (const auto & p : paths_) {
-    path.points.insert(path.points.end(), p.points.begin(), p.points.end());
+  for (const auto & partial_path : paths_) {
+    path.points.insert(path.points.end(), partial_path.points.begin(), partial_path.points.end());
   }
-  return path;
+
+  return removeOverlappingPoints(path);
 }
 
 PathWithLaneId GeometricParallelParking::getArcPath() const
 {
   PathWithLaneId path{};
-  for (size_t i = 1; i < paths_.size(); i++) {
-    const auto p = paths_.at(i);
-    path.points.insert(path.points.end(), p.points.begin(), p.points.end());
+  for (const auto & arc_path : arc_paths_) {
+    path.points.insert(path.points.end(), arc_path.points.begin(), arc_path.points.end());
   }
   return path;
 }
 
-void GeometricParallelParking::clear()
+bool GeometricParallelParking::isParking() const { return current_path_idx_ > 0; }
+
+bool GeometricParallelParking::isEnoughDistanceToStart(const Pose & start_pose) const
+{
+  const Pose current_pose = planner_data_->self_pose->pose;
+  const Pose current_to_start =
+    inverseTransformPose(start_pose, current_pose);  // todo: arc length is better
+
+  // not enough to stop with max deceleration
+  const double current_vel = util::l2Norm(planner_data_->self_odometry->twist.twist.linear);
+  const double stop_distance = std::pow(current_vel, 2) / parameters_.maximum_deceleration / 2;
+  if (current_to_start.position.x < stop_distance) {
+    return false;
+  }
+
+  // not enough to restart from stopped
+  constexpr double min_restart_distance = 3.0;
+  if (
+    current_vel < parameters_.th_stopped_velocity_mps &&
+    current_to_start.position.x > parameters_.th_arrived_distance_m &&
+    current_to_start.position.x < min_restart_distance) {
+    return false;
+  }
+
+  return true;
+}
+
+void GeometricParallelParking::setVelocityToArcPaths(
+  std::vector<PathWithLaneId> & arc_paths, const double velocity)
+{
+  for (auto & path : arc_paths) {
+    for (size_t i = 0; i < path.points.size(); i++) {
+      if (i == path.points.size() - 1) {
+        // stop point at the end of the path
+        path.points.at(i).point.longitudinal_velocity_mps = 0.0;
+      } else {
+        path.points.at(i).point.longitudinal_velocity_mps = velocity;
+      }
+    }
+  }
+}
+
+std::vector<PathWithLaneId> GeometricParallelParking::generateParkingPaths(
+  const Pose & start_pose, const Pose & goal_pose, const double R_E_r,
```

Comment:
what mean R_E_r?

Reply:
This comes from the referenced paper. This is the turning radius of the right turn of the base link.

https://github.com/tier4/autoware.universe/blob/docs/update_pull_over_docs/planning/behavior_path_planner/README.md#references--external-links

![image](https://user-images.githubusercontent.com/39142679/185523585-c74ef951-d693-464b-b67c-3ecc7ee84696.png)

Reply:
I see. can you add the reference to README?

Reply:
I have added it to the README of behavior_path as linked above.

Reply:
:+1: